node/src/components/linear_chain_sync/operations.rs

```diff
@@ -0,0 +1,288 @@
+/// in the trie-store, and return any outstanding descendant tries.
+async fn fetch_trie_and_insert_into_trie_store<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    trie_key: Blake2bHash,
+) -> Result<Vec<Blake2bHash>, LinearChainSyncError<I>>
+where
+    REv: From<ContractRuntimeRequest>
+        + From<FetcherRequest<I, Trie<Key, StoredValue>>>
+        + From<NetworkInfoRequest<I>>,
+    I: Eq + Clone + Debug + Send + 'static,
+{
+    let trie =
+        fetch_retry_forever::<Trie<Key, StoredValue>, REv, I>(effect_builder, trie_key).await?;
+    let outstanding_tries = effect_builder
+        .put_trie_and_find_missing_descendant_trie_keys(trie)
+        .await?;
+    return Ok(outstanding_tries);
+}
+
+/// Run the fast synchronization task.
+pub(crate) async fn run_fast_sync_task<REv, I>(
+    effect_builder: EffectBuilder<REv>,
+    trusted_hash: BlockHash,
+    chainspec: Chainspec,
+) -> Result<(), LinearChainSyncError<I>>
+where
+    REv: From<ContractRuntimeRequest>
+        + From<FetcherRequest<I, BlockHeader>>
+        + From<FetcherRequest<I, BlockHeaderWithMetadata>>
+        + From<FetcherRequest<I, Trie<Key, StoredValue>>>
+        + From<NetworkInfoRequest<I>>,
+    I: Eq + Debug + Clone + Send + 'static,
+{
+    // Fetch the trusted header
+    // TODO: save header after successful fetch
+    let trusted_header =
+        fetch_retry_forever::<BlockHeader, REv, I>(effect_builder, trusted_hash).await?;
+
+    // Walk back to get validators from the last switch block header
+    // First, check that we are at least one era past genesis or the last emergency restart
+    let maybe_last_emergency_restart_era_id = chainspec.protocol_config.last_emergency_restart;
+    let min_era = maybe_last_emergency_restart_era_id.unwrap_or_else(|| EraId::new(0));
+    if min_era >= trusted_header.era_id() {
+        return Err(LinearChainSyncError::TrustedHeaderEraTooEarly {
+            trusted_header,
+            maybe_last_emergency_restart_era_id,
+        });
+    };
+
+    // Fetch each parent hash one by one until we have the trusted validator weights
+    let mut current_walk_back_header = trusted_header.clone();
+    let mut trusted_validator_weights = loop {
+        if let Some(trusted_validator_weights) =
+            current_walk_back_header.next_era_validator_weights()
+        {
+            break trusted_validator_weights.clone();
+        }
+        current_walk_back_header = fetch_retry_forever::<BlockHeader, REv, I>(
+            effect_builder,
+            *current_walk_back_header.parent_hash(),
+        )
+        .await?;
+    };
+
+    // Get the most recent header which has the same version as ours
+    // We keep fetching by height until none of our peers have a block at that height
+    let mut most_recent_block_header = *trusted_header;
+    loop {
+        // If we encounter a block header of a version which is newer than ours we return an error
```

Comment:
Will this result in a shutdown with exit code 0, as in [CEP 37](https://github.com/casper-network/ceps/pull/37)?

Reply:
Yes!  The error here is ultimately handled as a `fatal` effect in `linear_chain_sync.rs`